### PR TITLE
Added `isNotEmpty` to collections

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes https://github.com/javaslang/javaslang/issues/1316

### CODE REVIEW

File: javaslang/src/main/java/javaslang/Value.java [hunk 1]
```diff
@@ -328,9 +328,9 @@ default T getOrElseTry(Try.CheckedSupplier<? extends T> supplier) {
     }
 
     /**
-     * Checks, this {@code Value} is empty, i.e. if the underlying value is absent.
+     * Checks whether this {@code Value} is empty, i.e. if the underlying value is absent.
      *
-     * @return false, if no underlying value is present, true otherwise.
+     * @return true, if no underlying value is present, false otherwise.
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/Value.java]: I think it was reversed

File: javaslang/src/main/java/javaslang/collection/Stack.java [hunk 2]
```diff
@@ -26,14 +26,14 @@
...
- *
+ * <p>
  * See Okasaki, Chris: <em>Purely Functional Data Structures</em> (p. 7 ff.). Cambridge, 2003.
  *
  * @param <T> component type
  * @author Daniel Dietrich
  * @since 2.0.0
  */
-public interface Stack<T> {
+public interface Stack<T> extends Traversable<T> {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Stack.java]: @danieldietrich, is this ok? We need to inherit the default `isNotNull`...

File: javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkResultAggregator.java [hunk 3]
```diff
@@ -18,7 +18,7 @@ public static void displayRatios(Collection<RunResult> runResultsCollection) {
             for (Tuple2<String, Array<RunResult>> group : getGroupsBasedOnBenchmarkPrefix(runResults)) {
                 final Multimap<Integer, Tuple2<String, Double>> results = aggregate(group._2);
                 final Multimap<String, Double> ratios = getRatios(results);
-                if (!ratios.isEmpty()) {
+                if (ratios.isNotEmpty()) {
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkResultAggregator.java]: changed all the code references to take advantage of the new API method

File: javaslang/src/main/java/javaslang/collection/RedBlackTree.java [hunk 4]
```diff
@@ -496,10 +505,10 @@ private boolean isLeaf() {
         private static <T> Node<T> balanceLeft(Color color, int blackHeight, RedBlackTree<T> left, T value,
                                                RedBlackTree<T> right, Empty<T> empty) {
             if (color == BLACK) {
-                if (!left.isEmpty()) {
+                if (left.isNotEmpty()) {
                     final Node<T> ln = (Node<T>) left;
                     if (ln.color == RED) {
-                        if (!ln.left.isEmpty()) {
+                        if (ln.left.isNotEmpty()) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/RedBlackTree.java]: nice example of why I recommended introducing this method

File: javaslang/src/main/java/javaslang/Value.java (hunk 1 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/Value.java]: yep

File: javaslang/src/main/java/javaslang/collection/Stack.java (hunk 2 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Stack.java]: We should not extend Traversable. If we need the method, it has to be duplicated (without inheritance).

There was a long discussion some time ago whether we should remove the Stack interface or not. We decided to keep it as rudimentary standalone interface. It is not a collection but a contract for Stack-like collections. List is currently the only implementation.

File: javaslang/src/main/java/javaslang/collection/List.java [hunk 5]
```diff
@@ -941,7 +941,7 @@ default T peek() {
         final Deque<T> preceding = new ArrayDeque<>(size());
         List<T> result = this;
         boolean found = false;
-        while (!found && !result.isEmpty()) {
+        while (!found && result.isNotEmpty()) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/List.java]: Generally I'm used to see `!someVar` or `!someMethodCall` and I would not have changed all occurreces of `!isEmpty` to `isNotEmpty`. `!` is part of the Java language and it is used everywhere. We can't live without it.

This line is a good example. We still have `!found`. We could rename it `notFound` and invert the content but that seems too complicated. We had to rewrite well-known algorithms to fit the new notation. Also it is not possible to eliminate `!` in algorithms that use both `someVar` and `!someVar`. So it is the best to accept `!`.

For the health of Javaslang it is better not to apply highly opinionated changes all over the code repository (a change may invent errors). It is like formatting. Some like 2 spaces indentation, some like 4 spaces or even tabs. Some prefer to use `!someVar`, some not.

Adding the additional method to `HashArrayMappedTrie` (HAMT) makes not much sense. We add code to an internal class just for a flavor of readability.

--

With this change we should add the method `isNotEmpty` to `Value` in order to offer the users of the _public_ API different ways to express themselves. But we should not change the whole code-base.

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/List.java]: @paplorinc sorry for having ruined this PR! You are very fast with refactorings in the broad - it is hard for me to follow and spend enough time to think about changes. Keep up the awesome work! You already have made significant changes to Javaslang!

**Pull request closed without merging**